SOSS/dms/soss_solver.py

```diff
@@ -0,0 +1,302 @@
+
+    :param simple_transform: The transformation parameters returned by
+        solve_transform().
+    :param ref_map: A reference map, e.g. a 2D Wavelength map or Trace Profile
+        map.
+    :param oversample: The oversampling factor the reference map.
+    :param pad: The padding (in native pixels) on the reference map.
+    :param native: If True bin down to native pixel sizes and remove padding.
+        Default is True
+    :param norm: If True normalize columns to 1, used for trace profile
+        reference maps. Default is False.
+
+    :type simple_transform: Tuple, List, Array
+    :type ref_map: array[float]
+    :type oversample: int
+    :type pad: int
+    :type native: bool
+    :type norm: bool
+
+    :returns: trans_maps - the ref_maps after having the transformation applied.
+    :rtype: array[float]
+    """
+
+    ovs = oversample
+
+    # Unpack the transformation.
+    angle, xshift, yshift = simple_transform
+
+    # Modify the transformation with the oversampling and padding.
+    xshift = ovs*xshift
+    yshift = ovs*yshift
+    cenx = ovs*(pad + 1024)
+    ceny = ovs*(pad + 50)
+
+    # Set NaN pixels to zero - the rotation doesn't handle NaNs well.
+    ref_map[np.isnan(ref_map)] = 0
+
+    # Apply the transformation to the reference map.
+    trans_map = transform_image(-angle, xshift, yshift, ref_map, cenx, ceny)
+
+    if native:
+        # Bin the transformed map down to native resolution.
+        nrows, ncols = trans_map.shape
+        trans_map = trans_map.reshape(nrows//ovs, ovs, ncols//ovs, ovs)
+        trans_map = trans_map.mean(1).mean(-1)
+
+        # Remove the padding.
+        trans_map = trans_map[pad:-pad, pad:-pad]
+
+    if norm:
+        # Normalize so that the columns sum to 1.
+        with warnings.catch_warnings():
+            warnings.simplefilter(action="ignore", category=RuntimeWarning)
+            trans_map = trans_map/np.nansum(trans_map, axis=0)
+
+    return trans_map
+
+
+def write_to_file(stack, filename):  # TODO function not needed?
+    """Utility function to write transformed 2D trace profile or wavelength map
+    files to disk. Data will be saved as a multi-extension fits file.
+
+    Parameters
+    ----------
+    stack : np.ndarray (2xYx2048)
+        Array containing transformed 2D trace profile or wavelength map data.
+        The first dimension must be the spectral order, the second dimension
+        the spatial dimension, and the third the spectral dimension.
+    filename : str
+        Name of the file to which to write the data.
+    """
+
+    hdu_p = fits.PrimaryHDU()
+    hdulist = [hdu_p]
+    for order in [1, 2]:
+        hdu_o = fits.ImageHDU(data=stack[order-1])
+        hdu_o.header['ORDER'] = order
+        hdu_o.header.comments['ORDER'] = 'Spectral order.'
+        hdulist.append(hdu_o)
+
+    hdu = fits.HDUList(hdulist)
+    hdu.writeto('{}.fits'.format(filename), overwrite=True)
```

Comment:
I know there is a TODO that this isn't needed but you should close the `hdu` once finished with it - can avoid memory leaks - just a note if this is to be kept in

Reply:
fixed in commit 8feda64